### PR TITLE
refactor: refactor station graph to isolate algorithm, data structure and topological

### PR DESCRIPTION
**Pull request by CodeBear801**
# Issue
https://github.com/Telenav/osrm-backend/issues/282

## Description              
- Remove `neighbor` in `node`, keep connectivity information inside `graph`
- Abstract the layer of interface `Graph`, make algorithm such as `Dijkstra` depends on interface only
- Abstract the layer of `connectivitymap.Querier`, which hides the logic for how connectivity is been built
   + To support `search along route charge station query`, user just pass in the connectivity graph of charge stations between low energy point
   + To support `charge station based routing`, user just pass in the pre-built connectivity graph between each charge stations
    + either of them could be implemented by OSRM cloud query or pre-built base google:s2+OSRM
- Remove logic of building graph in `stationgraph`, put them inside of `node_graph`.  All connectivity is build on the fly, not like previous version, `stationgraph` babysits the creation of graph.  The responsibility of `stationgraph` now is calling internal functions to do the work and convert result to certain format.
- More information could be found here: https://github.com/Telenav/osrm-backend/blob/902202640fc6ada64a7dae3dd4eee83a431814ae/integration/service/oasis/stationgraph/doc.go#L1

- To to:
    + later will consider change name of `connectivitymap.Querier` to `connectivitymap.TopoQuerier`, still considering
    + remove `locationinfo`, use `nav.location`
    + put logic into `internal`

## Tasklist

 - [x] make changes
 - [x] add tests

### CODE REVIEW

**Comment by wangyoucao577** [comment]
Will read it later. Please follow the PR template and always add entry in CHANGELOG-FORK.md. The rules in https://github.com/Telenav/osrm-backend/wiki/CHANGELOG.

File: integration/service/oasis/stationgraph/edge.go [hunk 1]
```diff
@@ -1,7 +1,16 @@
...
+	fromNodeID nodeID
+	toNodeID   nodeID
+}
+
+type edge struct {
+	distance float64
+	duration float64
+}
+
+type edgeIDAndData struct {
```
Comment (wangyoucao577) [inline integration/service/oasis/stationgraph/edge.go:13]: I'll prefer to name it `edge`. Name the the former `edge` to `metric` or something like this.

File: integration/service/oasis/stationgraph/node_container.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		nc.counter++
+
+		return n
+	}
+}
+
+func (nc *nodeContainer) getNode(id nodeID) *node {
+	if n, ok := nc.id2NodePtr[id]; ok {
+		return n
+	} else {
```
Comment (wangyoucao577) [inline integration/service/oasis/stationgraph/node_container.go:52]: Prefer to remove the `else`. See https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Comment (CodeBear801) [reply integration/service/oasis/stationgraph/node_container.go:52]: agree.

File: integration/service/oasis/stationgraph/edge.go (hunk 1 above)
Comment (CodeBear801) [reply integration/service/oasis/stationgraph/edge.go:13]: Good suggestion.  Will consider to make the change in the next pull.

**Pull request merged**